Support sideways-lr coordinate conversion for carets and selection

fantasai opened the pull request:
#### 12e65d787eb09add291c21e11858bf6d4d7bef5c
<pre>
Support sideways-lr coordinate conversion for carets and selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=286961">https://bugs.webkit.org/show_bug.cgi?id=286961</a>
<a href="https://rdar.apple.com/144117372">rdar://144117372</a>

Reviewed by Alan Baradlay.

In sideways-lr mode (uniquely), LTR text progresses in the direction opposite
to the direction our inline coordinates increase. Update caret and selection
coordinate computations, comparisons, and box traversals to handle this mismatch.

This commit introduces two cosmetic changes for consistency with introduced code:
- Drop the style() in formattingContext().style().writingMode() in two places
- Rename pointLineDirection and pointBlockDirection to logicalPoint.x()/y()

* LayoutTests/editing/caret/caret-position-sideways-lr-expected.txt: Added.
* LayoutTests/editing/caret/caret-position-sideways-lr.html: Added.
* LayoutTests/platform/ios/editing/caret/caret-position-sideways-lr-expected.txt: Added.
* LayoutTests/platform/mac-ventura/editing/caret/caret-position-sideways-lr-expected.txt: Added.
* LayoutTests/platform/mac/editing/caret/caret-position-sideways-lr-expected.txt: Added.
* LayoutTests/platform/win/editing/caret/caret-position-sideways-lr-expected.txt: Added.
* Source/WebCore/layout/integration/inline/InlineIteratorBox.h:
(WebCore::InlineIterator::Box::isInlineFlipped const): Introduce new convenience method.
* Source/WebCore/layout/integration/inline/InlineIteratorBoxInlines.h:
(WebCore::InlineIterator::Box::logicalLeft const): Provide Render tree logical coords.
(WebCore::InlineIterator::Box::logicalRight const): Provide Render tree logical coords.
(WebCore::InlineIterator::LeafBoxIterator::traverseLogicalRightwardOnLine): Add coordinate-relative traversal methods.
(WebCore::InlineIterator::LeafBoxIterator::traverseLogicalLeftwardOnLine): Add coordinate-relative traversal methods.
(WebCore::InlineIterator::LeafBoxIterator::traverseLogicalRightwardOnLineIgnoringLineBreak): Add coordinate-relative traversal methods.
(WebCore::InlineIterator::LeafBoxIterator::traverseLogicalLeftwardOnLineIgnoringLineBreak): Add coordinate-relative traversal methods.
* Source/WebCore/layout/integration/inline/InlineIteratorLineBox.cpp:
(WebCore::InlineIterator::closestBoxForHorizontalPosition): Update to coordinate flow traversal.
* Source/WebCore/layout/integration/inline/InlineIteratorLineBox.h:
* Source/WebCore/layout/integration/inline/InlineIteratorLineBoxInlines.h:
(WebCore::InlineIterator::LineBox::logicalLeftmostLeafBox const): Add coordinate-relative traversal methods.
(WebCore::InlineIterator::LineBox::logicalRightmostLeafBox const): Add coordinate-relative traversal methods.
* Source/WebCore/layout/integration/inline/InlineIteratorLineBoxModernPath.h:
(WebCore::InlineIterator::LineBoxIteratorModernPath::contentLogicalTopAdjustedForPrecedingLineBox const): Tweak for consistency.
(WebCore::InlineIterator::LineBoxIteratorModernPath::contentLogicalBottomAdjustedForFollowingLineBox const): Tweak for consistency.
(WebCore::InlineIterator::LineBoxIteratorModernPath::contentLogicalLeft const): Add support for sideways-lr coordinates.
* Source/WebCore/rendering/CaretRectComputation.cpp:
(WebCore::computeCaretRectForText): Separate out getting the textRun offset from adding the box edge offset, and perform coordinate conversion on the former only.
(WebCore::computeCaretRectForEmptyElement): Update caret alignment to support sideways-lr.
(WebCore::computeCaretRectForLinePosition): Update caret alignment to support sideways-lr.
(WebCore::computeCaretRectForLineBreak):
(WebCore::computeCaretRectForSVGInlineText): Incorporate sideways-lr logic into inline direction check.
(WebCore::computeCaretRectForBox): Incorporate sideways-lr logic into inline direction check.
* Source/WebCore/rendering/LegacyInlineTextBox.cpp:
(WebCore::LegacyInlineTextBox::localSelectionRect const): Separate out getting the textRun offset from adding the box edge offset, and perform coordinate conversion on the former only.
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::selectionRectForTextBox): Separate out getting the textRun offset from adding the box edge offset, and perform coordinate conversion on the former only.
(WebCore::offsetForPositionInRun): Incorporate sidways-lr logic into inline direction checks, and convert sideways-lr coordinates appropriately.
(WebCore::lineDirectionPointFitsInBox): Perform coordinate comparisons in the correct coordinate system, using coordinate-relative traversal methods.
(WebCore::RenderText::positionForPoint): Perform coordinate comparisons in the correct coordinate system. Rename local variables accordingly.

Canonical link: <a href="https://commits.webkit.org/290387@main">https://commits.webkit.org/290387@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/76bfee0f20705e1c3025355e9aaa63a2d33ed9e2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89702 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9229 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44578 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94695 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40469 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91754 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9618 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17508 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69084 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26708 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92703 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7395 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81392 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49450 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7103 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35830 "Found 1 new test failure: media/modern-media-controls/invalid-placard/invalid-placard-constrained-metrics.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39598 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77442 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36801 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96522 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16884 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12382 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77950 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17140 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77214 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77279 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21710 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20283 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10054 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14113 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16897 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22213 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16638 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20089 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18420 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay reviewed (approved):
I have to trust our test coverage here as it's really difficult to figure out what's going.

fantasai committed:
Support sideways-lr coordinate conversion for carets and selection
https://bugs.webkit.org/show_bug.cgi?id=286961
rdar://144117372

Reviewed by Alan Baradlay.

In sideways-lr mode (uniquely), LTR text progresses in the direction opposite
to the direction our inline coordinates increase. Update caret and selection
coordinate computations, comparisons, and box traversals to handle this mismatch.

This commit introduces two cosmetic changes for consistency with introduced code:
- Drop the style() in formattingContext().style().writingMode() in two places
- Rename pointLineDirection and pointBlockDirection to logicalPoint.x()/y()

* LayoutTests/editing/caret/caret-position-sideways-lr-expected.txt: Added.
* LayoutTests/editing/caret/caret-position-sideways-lr.html: Added.
* LayoutTests/platform/ios/editing/caret/caret-position-sideways-lr-expected.txt: Added.
* LayoutTests/platform/mac-ventura/editing/caret/caret-position-sideways-lr-expected.txt: Added.
* LayoutTests/platform/mac/editing/caret/caret-position-sideways-lr-expected.txt: Added.
* LayoutTests/platform/win/editing/caret/caret-position-sideways-lr-expected.txt: Added.
* Source/WebCore/layout/integration/inline/InlineIteratorBox.h:
(WebCore::InlineIterator::Box::isInlineFlipped const): Introduce new convenience method.
* Source/WebCore/layout/integration/inline/InlineIteratorBoxInlines.h:
(WebCore::InlineIterator::Box::logicalLeft const): Provide Render tree logical coords.
(WebCore::InlineIterator::Box::logicalRight const): Provide Render tree logical coords.
(WebCore::InlineIterator::LeafBoxIterator::traverseLogicalRightwardOnLine): Add coordinate-relative traversal methods.
(WebCore::InlineIterator::LeafBoxIterator::traverseLogicalLeftwardOnLine): Add coordinate-relative traversal methods.
(WebCore::InlineIterator::LeafBoxIterator::traverseLogicalRightwardOnLineIgnoringLineBreak): Add coordinate-relative traversal methods.
(WebCore::InlineIterator::LeafBoxIterator::traverseLogicalLeftwardOnLineIgnoringLineBreak): Add coordinate-relative traversal methods.
* Source/WebCore/layout/integration/inline/InlineIteratorLineBox.cpp:
(WebCore::InlineIterator::closestBoxForHorizontalPosition): Update to coordinate flow traversal.
* Source/WebCore/layout/integration/inline/InlineIteratorLineBox.h:
* Source/WebCore/layout/integration/inline/InlineIteratorLineBoxInlines.h:
(WebCore::InlineIterator::LineBox::logicalLeftmostLeafBox const): Add coordinate-relative traversal methods.
(WebCore::InlineIterator::LineBox::logicalRightmostLeafBox const): Add coordinate-relative traversal methods.
* Source/WebCore/layout/integration/inline/InlineIteratorLineBoxModernPath.h:
(WebCore::InlineIterator::LineBoxIteratorModernPath::contentLogicalTopAdjustedForPrecedingLineBox const): Tweak for consistency.
(WebCore::InlineIterator::LineBoxIteratorModernPath::contentLogicalBottomAdjustedForFollowingLineBox const): Tweak for consistency.
(WebCore::InlineIterator::LineBoxIteratorModernPath::contentLogicalLeft const): Add support for sideways-lr coordinates.
* Source/WebCore/rendering/CaretRectComputation.cpp:
(WebCore::computeCaretRectForText): Separate out getting the textRun offset from adding the box edge offset, and perform coordinate conversion on the former only.
(WebCore::computeCaretRectForEmptyElement): Update caret alignment to support sideways-lr.
(WebCore::computeCaretRectForLinePosition): Update caret alignment to support sideways-lr.
(WebCore::computeCaretRectForLineBreak):
(WebCore::computeCaretRectForSVGInlineText): Incorporate sideways-lr logic into inline direction check.
(WebCore::computeCaretRectForBox): Incorporate sideways-lr logic into inline direction check.
* Source/WebCore/rendering/LegacyInlineTextBox.cpp:
(WebCore::LegacyInlineTextBox::localSelectionRect const): Separate out getting the textRun offset from adding the box edge offset, and perform coordinate conversion on the former only.
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::selectionRectForTextBox): Separate out getting the textRun offset from adding the box edge offset, and perform coordinate conversion on the former only.
(WebCore::offsetForPositionInRun): Incorporate sidways-lr logic into inline direction checks, and convert sideways-lr coordinates appropriately.
(WebCore::lineDirectionPointFitsInBox): Perform coordinate comparisons in the correct coordinate system, using coordinate-relative traversal methods.
(WebCore::RenderText::positionForPoint): Perform coordinate comparisons in the correct coordinate system. Rename local variables accordingly.

Canonical link: https://commits.webkit.org/290387@main
